File: verify-sdk-boot/boot_check.rb
# frozen_string_literal: true
# Boot-check for sdk-ruby. See run.sh for context.

def fail_check(msg)
  warn "FAIL sdk-ruby: #{msg}"
  exit 1
end

fixture = ENV['QFG_BOOT_CHECK_FIXTURE_HOME']
fail_check 'QFG_BOOT_CHECK_FIXTURE_HOME unset' if fixture.nil? || fixture.empty?

# Dir.home reads $HOME first on Unix; point it at the fixture in-process
# so sdk-ruby's loader finds the synthetic tokens.json without disturbing
# the outer shell's HOME.
ENV['HOME'] = fixture

if ENV.fetch('QUONFIG_BACKEND_SDK_KEY', '') != ''
  fail_check 'QUONFIG_BACKEND_SDK_KEY must be unset — boot-check exists to prove the SDK boots without it'
end

datadir = ENV['QFG_BOOT_CHECK_DATADIR']
expected_email = ENV['QFG_BOOT_CHECK_EXPECTED_EMAIL']
expected_key = ENV['QFG_BOOT_CHECK_EXPECTED_KEY']
expected_value = ENV['QFG_BOOT_CHECK_EXPECTED_VALUE']
[datadir, expected_email, expected_key, expected_value].each_with_index do |v, i|
  fail_check "missing QFG_BOOT_CHECK_* env var (index #{i})" if v.nil? || v.empty?
end

$LOAD_PATH.unshift(File.expand_path('../../sdk-ruby/lib', __dir__))
begin
  require 'quonfig'
rescue LoadError => e
  fail_check "require 'quonfig' failed: #{e.class}: #{e.message}"
end

dev_ctx = Quonfig::DevContext.load_quonfig_user_context
fail_check 'dev_context loader returned nil — synthetic tokens.json not picked up' if dev_ctx.nil?
actual_email = dev_ctx.dig('quonfig-user', 'email')
unless actual_email == expected_email
  fail_check "loader returned quonfig-user.email=#{actual_email.inspect}, expected #{expected_email.inspect}"
end

begin
  opts = Quonfig::Options.new(
    datadir: datadir,
    environment: 'Production',
    enable_quonfig_user_context: true,
    collect_evaluation_summaries: false
  )
  client = Quonfig::Client.new(opts)
rescue StandardError => e
  fail_check "Quonfig::Client.new raised without sdk_key: #{e.class}: #{e.message}"
end

begin
  value = client.get_string(expected_key, default: '__BOOT_CHECK_DEFAULT__')
rescue StandardError => e
  fail_check "client.get_string raised: #{e.class}: #{e.message}"
end

unless value == expected_value
  fail_check "client.get_string(#{expected_key.inspect}) returned #{value.inspect}, expected #{expected_value.inspect}"
end

puts "OK sdk-ruby: constructed without sdk_key, dev_context email=#{expected_email}, " \
     "get_string(#{expected_key.inspect})=#{value.inspect}"

File: verify-sdk-boot/boot_check.mjs
// Boot-check for sdk-node. See run.sh for context.
import process from "node:process";
import { pathToFileURL } from "node:url";
import path from "node:path";

function fail(msg) {
  console.error(`FAIL sdk-node: ${msg}`);
  process.exit(1);
}

const fixture = process.env.QFG_BOOT_CHECK_FIXTURE_HOME;
if (!fixture) fail("QFG_BOOT_CHECK_FIXTURE_HOME unset");
// os.homedir() reads $HOME first on Unix; setting it in-process points
// the SDK's dev-context loader at our synthetic tokens.json without
// disturbing the outer shell's HOME (npm cache, etc.).
process.env.HOME = fixture;

if (process.env.QUONFIG_BACKEND_SDK_KEY) {
  fail("QUONFIG_BACKEND_SDK_KEY must be unset — boot-check exists to prove the SDK boots without it");
}

const datadir = process.env.QFG_BOOT_CHECK_DATADIR;
const expectedEmail = process.env.QFG_BOOT_CHECK_EXPECTED_EMAIL;
const expectedKey = process.env.QFG_BOOT_CHECK_EXPECTED_KEY;
const expectedValue = process.env.QFG_BOOT_CHECK_EXPECTED_VALUE;
if (!datadir || !expectedEmail || !expectedKey || !expectedValue) {
  fail("missing QFG_BOOT_CHECK_* env vars");
}

const here = path.dirname(new URL(import.meta.url).pathname);
const sdkEntry = path.resolve(here, "..", "..", "sdk-node", "dist", "index.js");
let mod;
try {
  mod = await import(pathToFileURL(sdkEntry).href);
} catch (e) {
  fail(`import @quonfig/node from ${sdkEntry}: ${e?.stack || e}`);
}
const { Quonfig } = mod;
if (!Quonfig) fail("sdk-node did not export Quonfig");

let q;
try {
  q = new Quonfig({
    datadir,
    environment: "Production",
    enableQuonfigUserContext: true,
    collectEvaluationSummaries: false,
    contextUploadMode: "none",
  });
} catch (e) {
  fail(`new Quonfig({}) raised without sdkKey: ${e?.stack || e}`);
}

try {
  await q.init();
} catch (e) {
  fail(`q.init() raised: ${e?.stack || e}`);
}

let value;
try {
  value = q.get(expectedKey, "__BOOT_CHECK_DEFAULT__");
} catch (e) {
  fail(`q.get(${expectedKey}) raised: ${e?.stack || e}`);
}
if (value !== expectedValue) {
  fail(`q.get(${expectedKey}) returned ${JSON.stringify(value)}, expected ${JSON.stringify(expectedValue)}`);
}

// sdk-node doesn't re-export loadQuonfigUserContext or expose a public
// global-context getter. Read the private `globalContext` field on the
// client. JS doesn't enforce TS `private`, so this works; the cost is
// that an internal rename of the field breaks this assertion — which is
// the desired regression signal (the check exists precisely to catch
// "dev-context wiring silently stopped working" regressions).
const gc = q.globalContext;
const actualEmail = gc?.["quonfig-user"]?.email;
if (actualEmail !== expectedEmail) {
  fail(
    `client.globalContext.quonfig-user.email = ${JSON.stringify(actualEmail)}, ` +
      `expected ${JSON.stringify(expectedEmail)} (full: ${JSON.stringify(gc)})`
  );
}

console.log(
  `OK sdk-node: constructed without sdkKey, globalContext.quonfig-user.email=${expectedEmail}, ` +
    `get(${JSON.stringify(expectedKey)})=${JSON.stringify(expectedValue)}`
);
